src/textures/bitmap.cpp

```diff
@@ -164,103 +174,133 @@ class BitmapTexture final : public Texture<Float, Spectrum> {
             Throw("Invalid wrap mode \"%s\", must be one of: \"repeat\", "
                   "\"mirror\", or \"clamp\"!", wrap_mode_str);
 
-        /* Convert to linear RGB float bitmap, will be converted
-           into spectral profile coefficients below (in place) */
-        Bitmap::PixelFormat pixel_format = m_bitmap->pixel_format();
-        switch (pixel_format) {
-            case Bitmap::PixelFormat::Y:
-            case Bitmap::PixelFormat::YA:
-                pixel_format = Bitmap::PixelFormat::Y;
-                break;
-
-            case Bitmap::PixelFormat::RGB:
-            case Bitmap::PixelFormat::RGBA:
-            case Bitmap::PixelFormat::XYZ:
-            case Bitmap::PixelFormat::XYZA:
-                pixel_format = Bitmap::PixelFormat::RGB;
-                break;
-
-            default:
-                Throw("The texture needs to have a known pixel "
-                      "format (Y[A], RGB[A], XYZ[A] are supported).");
-        }
-
         /* Should Mitsuba disable transformations to the stored color data?
            (e.g. sRGB to linear, spectral upsampling, etc.) */
         m_raw = props.get<bool>("raw", false);
-        if (m_raw) {
-            /* Don't undo gamma correction in the conversion below.
-               This is needed, e.g., for normal maps. */
-            m_bitmap->set_srgb_gamma(false);
-        }
-
         m_accel = props.get<bool>("accel", true);
 
-        // Convert the image into the working floating point representation
-        m_bitmap =
-            m_bitmap->convert(pixel_format, struct_type_v<ScalarFloat>, false);
+        if (tensor) {
+            Log(Debug, "Loading bitmap texture from tensor...");
+            if (!m_raw)
+                Throw("Bitmap raw property must be true when initializing using tensor data."
+                      "Otherwise use a bitmap object or file if transformation of color data is"
+                      " required.");
+            m_texture = Texture2f(TensorXf(*tensor), m_accel, m_accel, filter_mode, wrap_mode);
+            const size_t pixel_count = tensor->shape(1) * tensor->shape(0);
+            const size_t ch_count = tensor->shape(2);
+
+            if (ch_count == 3) {
+                if constexpr (dr::is_dynamic_v<Float>)
+                    m_mean = dr::sum(luminance(dr::unravel<Color3f>(tensor->array()))) / pixel_count;
+                else {
+                    const size_t pixel_count = tensor->shape(0) * tensor->shape(1);
+                    double mean = 0.0;
+                    ScalarFloat* ptr = tensor->data();
+                    for (size_t i = 0; i < pixel_count; ++i) {
+                        ScalarColor3f value = dr::load<ScalarColor3f>(ptr);
+                        mean += (double) luminance(value);
+                        ptr += 3;
+                    }
+                    m_mean = (ScalarFloat)(mean / pixel_count);
```

Comment:
We should probably support spectral modes as well here (same as what we do for the bitmap input).

Reply:
The sRGB gamma conversion and the spectral up-sampling are scalar-only processes. I'd strongly prefer if it was somewhat explicit that a potential GPU->CPU conversion is necessary. Especially when the workaround is a straightforward one-liner: `mi.Bitmap(tensor_variable)`. 

In addition, I feel like it is clearer to a normal user that the `TensorXf` initialization is a fast path to setting data without anymodification ever.

Reply:
I understand. But does this mean that in spectral mode, when passing a RGB image as a `TensorXf` or a `Bitmap`, it will produce different images?

Reply:
Here are all the cases:
```
(input type, color variant, raw?) => (expected semantic meaning of input data, processing step)

bitmap, rgb,      !raw  => (sRGB values or linear RGB, gamma correction)
tensor, rgb,      !raw  => not supported

bitmap, rgb,       raw  => (linear RGB, copied)
tensor, rgb,       raw  => (linear RGB, copied)

bitmap, spectral, !raw  => (sRGB values or linear RGB, spectral upsampling)
tensor, spectral, !raw  => not supported

bitmap, spectral,  raw  => (spectral upsampling coefficients, copied)
tensor, spectral,  raw  => (spectral upsampling coefficients, copied)
```
So if you have linear RGB values in a tensor, in a spectral variant, you will need to do:
`'bitmap': mi.Bitmap(tensor), 'raw': False`
And if you have sRGB values in a tensor, in a spectral variant, you will need to do:
```python
bitmap = mi.Bitmap(tensor); bitmap.set_srgb_gamma(True);
(...)
'bitmap': bitmap, 'raw': False
```

This is my current understanding of what will happen. Am I correct @rtabbara ?

I think I'm happy with this, any issues you see @Speierers? We're missing metadata about the tensor to be smart about sRGB conversions, there's no perfect solution IMHO.

Reply:
Ok, that looks good to me. Thanks for the detailed table.

As long as we don't get different results, I am happy with the current solution. Here it seems that it either matches what `Bitmap` would do, or it is explicitly not supported. 👍